git/src/lib.rs

```diff
@@ -0,0 +1,206 @@
+    /// abbreviated hash it fetches everything and then looks for the hash among the
+    /// things that it finds.
+    Commit(#[serde_as(as = "serde_with::DisplayFromStr")] ObjectId),
+}
+
+impl std::fmt::Display for Target {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Target::Head => write!(f, "HEAD"),
+            Target::Branch(branch) => write!(f, "refs/heads/{branch}"),
+            Target::Tag(tag) => write!(f, "refs/tags/{tag}"),
+            Target::Commit(c) => write!(f, "{}", c),
+        }
+    }
+}
+
+impl Target {
+    fn refspec(&self) -> String {
+        self.to_string()
+    }
+}
+
+fn source_object_id(source: &gix::remote::fetch::Source) -> Result<ObjectId> {
+    match source {
+        gix::remote::fetch::Source::ObjectId(id) => Ok(*id),
+        gix::remote::fetch::Source::Ref(r) => {
+            let (_name, id, peeled) = r.unpack();
+
+            Ok(peeled
+                .or(id)
+                .ok_or_else(|| anyhow!("unborn reference"))?
+                .to_owned())
+        }
+    }
+}
+
+/// Fetches the contents of a git repository into directory.
+///
+/// The directory will be created if it doesn't exist yet. The data will be
+/// fetched from scratch, even if it has already been fetched before. However,
+/// we will try to minimize the amount of data to fetch (for example, by doing a
+/// shallow fetch).
+///
+/// Only the contents of the git repository will be written to the given
+/// directory; the git directory itself will be discarded.
+pub fn fetch(spec: &Spec, dir: impl AsRef<Path>) -> Result<ObjectId> {
+    let dir = dir.as_ref();
+    std::fs::create_dir_all(dir).with_path(dir)?;
+
+    // Fetch the git directory somewhere temporary.
+    let git_tempdir = tempfile::tempdir().wrap_err()?;
+    let repo = gix::init(git_tempdir.path()).wrap_err()?;
+    let refspec = spec.target.refspec();
+
+    let remote = repo
+        .remote_at(spec.url.clone())
+        .wrap_err()?
+        .with_fetch_tags(gix::remote::fetch::Tags::None)
+        .with_refspecs(Some(refspec.as_str()), gix::remote::Direction::Fetch)
+        .wrap_err()?;
+
+    // This does similar credentials stuff to the git CLI (e.g. it looks for ssh
+    // keys if it's a fetch over ssh, or it tries to run `askpass` if it needs
+    // credentials for https). Maybe we want to have explicit credentials
+    // configuration instead of or in addition to the default?
+    let connection = remote.connect(gix::remote::Direction::Fetch).wrap_err()?;
+    let outcome = connection
+        .prepare_fetch(
+            &mut gix::progress::Discard,
+            gix::remote::ref_map::Options::default(),
+        )
+        .wrap_err()?
+        // For now, we always fetch shallow. Maybe for the index it's more efficient to
+        // keep a single repo around and update it? But that might be in another method.
+        .with_shallow(gix::remote::fetch::Shallow::DepthAtRemote(
```

Comment:
Nitpick: there are a lot of `gix::remote::fetch::` prefixes. Maybe it would be worth importing at least `fetch`, or even a few symbols directly?